Adding NET 4.8 WinForms Accessibility changes compatibility document.

M-Lipin opened the pull request:
Adding **winforms-accessibility-changes-48.md** document which contains the description of accessibility changes introduced in .NET 4.8:
* DataGridView control Accessibility changes;
* Improvements for screen readers (Narrator);
* CheckedListBox control Accessibility changes;
* ComboBox control Accessibility changes;
* LinkLabel control Accessibility changes;
* ProgressBar control Accessibility changes;
* PropertyGrid control Accessibility changes;
* Tool Strip control Accessibility changes;
* Visual cues improvements;

**Note:** the document is not final version and should be updated with the changes related to Keyboard ToolTip, High contrast changes and other accessibility improvements introduced by WinForms team for .NET 4.8.

//cc @PriyaPurkayastha

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/dotnet?pullRequest=969) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/dotnet?pullRequest=969) before we can accept your contribution.<br/><br/>:x: M-Lipin [sign now](https://cla.opensource.microsoft.com/Microsoft/dotnet?pullRequest=969)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/dotnet?pullRequest=969) it.</sub>

Tanya-Solyanik commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+### Recommended Action
+__How to opt in or out of these changes__
+  
+In order for the application to benefit from these changes, it must run on the .NET Framework 4.8. The application can benefit from these changes in either of the following ways:
```
```suggestion
In order for the application to benefit from these changes, it must run on the .NET Framework 4.8. The application can opt-in into these changes in either of the following ways:
```

Tanya-Solyanik commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+- Improved <xref:System.Windows.Forms.LinkLabel> control accessibility: Narrator announces the disabled state for the link if corresponding <xref:System.Windows.Forms.LinkLabel> control is disabled.
+
+__Improved Progress Bar Accessibility support__
+- Enabled UI Automation support for <xref:System.Windows.Forms.ProgressBar> control with ability to use UI Automation notifications and other UI Automation features. Developers are now able to use UI Automation notifications which Narrator can announce to indicate the progress.
```
@M-Lipin  - is there a link to UIA notification you could add here?

M-Lipin replied:
Thanks for the comment!
I've added the link in 15c3da6c5585bdd314b3b75a285626a016c5d011.

Tanya-Solyanik commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+- Enabled UI Automation support for <xref:System.Windows.Forms.DataGridView> control with ability to use UI Automation notifications and other UI Automation features.
+- The UI Automation element which corresponds to the <xref:System.Windows.Forms.DataGridViewComboBoxEditingControl> or <xref:System.Windows.Forms.DataGridViewTextBoxEditingControl> is now a child of corresponding editing cell.
+
+__Improved Link Label Accessibility support__
```
```suggestion
__Improved LinkLabel Accessibility support__
```

M-Lipin replied:
Thanks!

Tanya-Solyanik commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+__Improved Link Label Accessibility support__
+- Improved <xref:System.Windows.Forms.LinkLabel> control accessibility: Narrator announces the disabled state for the link if corresponding <xref:System.Windows.Forms.LinkLabel> control is disabled.
+
+__Improved Progress Bar Accessibility support__
```
```suggestion
__Improved ProgressBar Accessibility support__
```

Tanya-Solyanik commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+- Enabled UI Automation support for <xref:System.Windows.Forms.ComboBox> control with ability to use UI Automation notifications and other UI Automation features.
+
+__Improved DataGridView Accessibility support__
+- Enabled UI Automation support for <xref:System.Windows.Forms.DataGridView> control with ability to use UI Automation notifications and other UI Automation features.
```
```suggestion
- Enabled UI Automation support for <xref:System.Windows.Forms.DataGridView> control with ability to use UI Automation notifications, other UI Automation features, and keyboard navigation of the control.
```
Though I'm not 100% sure, please verify, I think we added f3 for sort order access

M-Lipin replied:
I looked at this and found out that F3-sorting was added in 4.7.2.
.

Tanya-Solyanik commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+__Improved Progress Bar Accessibility support__
+- Enabled UI Automation support for <xref:System.Windows.Forms.ProgressBar> control with ability to use UI Automation notifications and other UI Automation features. Developers are now able to use UI Automation notifications which Narrator can announce to indicate the progress.
+
+__Improved Property Grid Accessibility support__
```
```suggestion
__Improved PropertyGrid Accessibility support__
```

Tanya-Solyanik commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+
+__Improved Property Grid Accessibility support__
+- Enabled UI Automation support for <xref:System.Windows.Forms.PropertyGrid> control with ability to use UI Automation notifications and other UI Automation features.
+- The UI Automation element which corresponds to the currently editing property is now a child of corresponding property item UI Automation element.
```
```suggestion
- The UI Automation element which corresponds to the currently edited property is now a child of the corresponding property item UI Automation element.
```

M-Lipin replied:
Thanks!

Tanya-Solyanik commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+__Improved Property Grid Accessibility support__
+- Enabled UI Automation support for <xref:System.Windows.Forms.PropertyGrid> control with ability to use UI Automation notifications and other UI Automation features.
+- The UI Automation element which corresponds to the currently editing property is now a child of corresponding property item UI Automation element.
+- The UI Automation property item element is now a child of corresponding category if the parent <xref:System.Windows.Forms.PropertyGrid> control is set to Category view.
```
```suggestion
- The UI Automation property item element is now a child of the corresponding category element if the parent <xref:System.Windows.Forms.PropertyGrid> control is set to category view.
```
I'm not sure about capitalization of the word `category`.

M-Lipin replied:
Thanks. I agree, category should not have 'c' capitalized.

Tanya-Solyanik commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+- The UI Automation element which corresponds to the currently editing property is now a child of corresponding property item UI Automation element.
+- The UI Automation property item element is now a child of corresponding category if the parent <xref:System.Windows.Forms.PropertyGrid> control is set to Category view.
+
+__Improved Tool Strip support__
```
```suggestion
__Improved ToolStrip support__
```

Tanya-Solyanik commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+- In items mode Narrator focus does not disappear and does not go to hidden items.
+
+__Improved Visual cues__
+- An empty <xref:System.Windows.Forms.CheckedListBox> control now displays a focus indicator when receives focus.
```
```suggestion
- An empty <xref:System.Windows.Forms.CheckedListBox> control now displays a focus indicator when it receives focus.
```

rpetrusha commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+NotPlanned
+
+### Change Description
+Windows Forms Framework is improving how it works with accessibility technologies to better support Windows Forms customers. These include the following changes:
```
```suggestion
The Windows Forms Framework is continuing to improve how it works with accessibility technologies to better support Windows Forms customers. These include the following changes:
```

rpetrusha commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+Windows Forms Framework is improving how it works with accessibility technologies to better support Windows Forms customers. These include the following changes:
+- Changes to improve display during High Contrast mode.
+- Changes to interaction with Narrator.
+- Changes in Accessible hierarchy. (improving navigation through UI Automation tree)
```
```suggestion
- Changes in the Accessible hierarchy (improving navigation through the UI Automation tree).
```

rpetrusha commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+  
+In order for the application to benefit from these changes, it must run on the .NET Framework 4.8. The application can benefit from these changes in either of the following ways:
+- It is recompiled to target the .NET Framework 4.8. These accessibility changes are enabled by default on Windows Forms applications that target the .NET Framework 4.8.
+- It targets the .NET Framework 4.7.2 or earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext Switch](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element) to the `<runtime>` section of the app config file and setting it to `false`, as the following example shows.
```
```suggestion
- It targets the .NET Framework 4.7.2 or earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext switch](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element) to the `<runtime>` section of the app config file and setting it to `false`, as the following example shows.
```

rpetrusha commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+- Narrator now announces the sort direction of the <xref:System.Windows.Forms.DataGridViewColumn> when announcing an accessible name of a <xref:System.Windows.Forms.DataGridViewCell>. 
+
+__Improved Checked ListBox Accessibility support__
+- Improved Narrator support for <xref:System.Windows.Forms.CheckedListBox> control. At this moment when navigating to the <xref:System.Windows.Forms.CheckedListBox> control using keyboard, Narrator focuses the <xref:System.Windows.Forms.CheckedListBox> item and announces it.
```
```suggestion
- Improved Narrator support for the <xref:System.Windows.Forms.CheckedListBox> control. When navigating to the <xref:System.Windows.Forms.CheckedListBox> control using the keyboard, Narrator focuses the <xref:System.Windows.Forms.CheckedListBox> item and announces it.
```

rpetrusha commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+- Improved Narrator support for <xref:System.Windows.Forms.CheckedListBox> control. At this moment when navigating to the <xref:System.Windows.Forms.CheckedListBox> control using keyboard, Narrator focuses the <xref:System.Windows.Forms.CheckedListBox> item and announces it.
+
+__Improved ComboBox Accessibility support__
+- Enabled UI Automation support for <xref:System.Windows.Forms.ComboBox> control with ability to use UI Automation notifications and other UI Automation features.
```
```suggestion
- Enabled UI Automation support for the <xref:System.Windows.Forms.ComboBox> control, with the ability to use UI Automation notifications and other UI Automation features.
```

rpetrusha commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+
+__Improved DataGridView Accessibility support__
+- Enabled UI Automation support for <xref:System.Windows.Forms.DataGridView> control with ability to use UI Automation notifications and other UI Automation features.
+- The UI Automation element which corresponds to the <xref:System.Windows.Forms.DataGridViewComboBoxEditingControl> or <xref:System.Windows.Forms.DataGridViewTextBoxEditingControl> is now a child of corresponding editing cell.
```
```suggestion
- The UI Automation element which corresponds to the <xref:System.Windows.Forms.DataGridViewComboBoxEditingControl> or <xref:System.Windows.Forms.DataGridViewTextBoxEditingControl> is now a child of the corresponding editing cell.
```

rpetrusha commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+- The UI Automation element which corresponds to the <xref:System.Windows.Forms.DataGridViewComboBoxEditingControl> or <xref:System.Windows.Forms.DataGridViewTextBoxEditingControl> is now a child of corresponding editing cell.
+
+__Improved Link Label Accessibility support__
+- Improved <xref:System.Windows.Forms.LinkLabel> control accessibility: Narrator announces the disabled state for the link if corresponding <xref:System.Windows.Forms.LinkLabel> control is disabled.
```
```suggestion
- Improved <xref:System.Windows.Forms.LinkLabel> control accessibility: Narrator announces the disabled state for the link if the corresponding <xref:System.Windows.Forms.LinkLabel> control is disabled.
```

rpetrusha commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+- Improved <xref:System.Windows.Forms.LinkLabel> control accessibility: Narrator announces the disabled state for the link if corresponding <xref:System.Windows.Forms.LinkLabel> control is disabled.
+
+__Improved Progress Bar Accessibility support__
+- Enabled UI Automation support for <xref:System.Windows.Forms.ProgressBar> control with ability to use UI Automation notifications and other UI Automation features. Developers are now able to use UI Automation notifications which Narrator can announce to indicate the progress.
```
```suggestion
- Enabled UI Automation support for the <xref:System.Windows.Forms.ProgressBar> control with the ability to use UI Automation notifications and other UI Automation features. Developers are now able to use UI Automation notifications which Narrator can announce to indicate progress.
```

rpetrusha commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+- Enabled UI Automation support for <xref:System.Windows.Forms.ProgressBar> control with ability to use UI Automation notifications and other UI Automation features. Developers are now able to use UI Automation notifications which Narrator can announce to indicate the progress.
+
+__Improved Property Grid Accessibility support__
+- Enabled UI Automation support for <xref:System.Windows.Forms.PropertyGrid> control with ability to use UI Automation notifications and other UI Automation features.
```
```suggestion
- Enabled UI Automation support for the <xref:System.Windows.Forms.PropertyGrid> control, with the ability to use UI Automation notifications and other UI Automation features.
```

rpetrusha commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+- The UI Automation property item element is now a child of corresponding category if the parent <xref:System.Windows.Forms.PropertyGrid> control is set to Category view.
+
+__Improved Tool Strip support__
+- Enabled UI Automation support for <xref:System.Windows.Forms.ToolStrip> control with ability to use UI Automation notifications and other UI Automation features.
```
```suggestion
- Enabled UI Automation support for the <xref:System.Windows.Forms.ToolStrip> control, with the ability to use UI Automation notifications and other UI Automation features.
```

rpetrusha commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,105 @@
+__Improved Tool Strip support__
+- Enabled UI Automation support for <xref:System.Windows.Forms.ToolStrip> control with ability to use UI Automation notifications and other UI Automation features.
+- Improved navigation through <xref:System.Windows.Forms.ToolStrip> items. 
+- In items mode Narrator focus does not disappear and does not go to hidden items.
```
```suggestion
- In items mode, Narrator focus does not disappear and does not go to hidden items.
```

bryukov commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,107 @@
+581351
+560840
+497307
+
```
```suggestion
548792
686499

```

bryukov commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,107 @@
+
+For an overview of UI automation, see the [UI Automation Overview](https://docs.microsoft.com/dotnet/framework/ui-automation/ui-automation-overview).</p>
+
+
```
```suggestion
__Invoking controls' ToolTips with a keyboard__
 Control tooltip can now be invoked by focusing the control with keyboard. 
 This feature needs to be enabled explicitly for the application (see section __"How to opt in or out of these changes"__)
 
```

bryukov commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
````diff
@@ -0,0 +1,107 @@
+```    
+Note that to opt in to the accessibility features added in .NET Framework 4.8, you must also opt in to accessibility features of .NET Framework 4.7.1 and 4.7.2 as well.
+Applications that target the .NET Framework 4.8 and want to preserve the legacy accessibility behavior can opt in to the use of legacy accessibility features by explicitly setting this AppContext switch to `true`.
+
````
```suggestion
Enabling the keyboard ToolTip invocation support requires adding the `Switch.System.Windows.Forms.UseLegacyToolTipDisplay=false` line to the AppContextSwitchOverrides value:
`<AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false;Switch.UseLegacyAccessibilityFeatures.3=false" />`

Note that enabling this feature requires opting in to the aforementioned accessibility features of .NET Framework 4.7.1 - 4.8.

```

M-Lipin commented:
Need to confirm that the following should (or should not) be added (not sure that these are in the email list but have 4.8 target):

Bug 544592: [Accessibility][Keyboard Navigation - BizTalk Server - Application - Send Port Gorups] Navigation is not cosistent when navigating in forward and backward direction on general window of Send port groups properties.
Change description: Improved ToolStrip keyboard navigation, made the ToolStrip forward and backward navigation consistent.

M-Lipin commented:
> Need to confirm that the following should (or should not) be added (not sure that these are in the email list but have 4.8 target):

Bug 544471: [Accessibility] DataGridView is reporting cells as always editable via Narrator.
Change description: made announcement for read-only DataGridView TextBox cell consistent: Narrator announces the cell as read-only.

M-Lipin commented:
> Need to confirm that the following should (or should not) be added (not sure that these are in the email list but have 4.8 target):

Bug 549288: Accessibility: WinForms ListView and DataGridView do not expose the current sort order through UIA.
Change description: The accessibility of DataGridView and ListView has been improved to make sort direction available via UIA: added exposing sort order and sort column via ItemStatus property and column name.

M-Lipin commented:
> Need to confirm that the following should (or should not) be added (not sure that these are in the email list but have 4.8 target):

Bug 599936: Accessibility:[Narrator]Narrator cannot annouce the ReadOnly Status for DataGridView TextBox column with setting Readonly as true.
Change description: The announcement of DataGridView Read-only TextBox cells is made accessible and clear: editable TextBoxes are announced as 'editable' and read-only TextBoxes are announced as 'read-only'.

M-Lipin commented:
> Need to confirm that the following should (or should not) be added (not sure that these are in the email list but have 4.8 target):

Bug 508369: Accessibility: [Narrator] Narrator is not announcing the updated value of ComboBox in PropertyGrid.
Change description: Added screen reader announcement for PropertyGrid's ComboBox value changes. Added ComboBox's text field update in response of ComboBox selection changes.

M-Lipin commented:
> Need to confirm that the following should (or should not) be added (not sure that these are in the email list but have 4.8 target):

Bug 615721: [Accessibility] The Invoke Pattern for DataGridViewCell: Image is not available.
Change description: Added support of Accessibility Invoke pattern to DataGridView Image cell with ability to invoke cell's default action. (in case Image cell is actually an image of a button)

M-Lipin commented:
> Need to confirm that the following should (or should not) be added (not sure that these are in the email list but have 4.8 target):

Bug 574309: Accessibility: MAS40B: Inspect: Expand/Collapse pattern property is FALSE for 'Type' & ‘Scope’ comboBoxes in 'Project properties node-->Settings page'.
Change description: Made the DataGridView's Combobox expand/collapse state accessible for the customers which use assistive technologies.

M-Lipin commented:
> Need to confirm that the following should (or should not) be added (not sure that these are in the email list but have 4.8 target):

Bug 564762: [Accessibility] The blue rectangle of selected ListView item should be keeping after tab or mouse to other control.
Change description: Improved ListView accessibility by enabling showing grey rectangle around the selected ListView item when parent ListView is out if focus.

rpetrusha commented:
Is this ready to merge, @M-Lipin?

M-Lipin commented:
> Is this ready to merge, @M-Lipin?

Hello @rpetrusha, I'll ask WinForms team for additional review, but basically it seems to me that this can be merged. Just need final confirmation.

rpetrusha commented:
OK. Thanks, @M-Lipin. Let me know when it can be merged.

Tanya-Solyanik commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,117 @@
+Note that to opt in to the accessibility features added in .NET Framework 4.8, you must also opt in to accessibility features of .NET Framework 4.7.1 and 4.7.2 as well.
+Applications that target the .NET Framework 4.8 and want to preserve the legacy accessibility behavior can opt in to the use of legacy accessibility features by explicitly setting this AppContext switch to `true`.
+Enabling the keyboard ToolTip invocation support requires adding the `Switch.System.Windows.Forms.UseLegacyToolTipDisplay=false` line to the AppContextSwitchOverrides value:
+`<AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false;Switch.UseLegacyAccessibilityFeatures.3=false" />`
```
```suggestion
`<AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false;Switch.UseLegacyAccessibilityFeatures.3=false;Switch.System.Windows.Forms.UseLegacyToolTIpDisplay=false" />`
```

Tanya-Solyanik commented on Documentation/compatibility/winforms-accessibility-changes-48.md:
```diff
@@ -0,0 +1,117 @@
+Enabling the keyboard ToolTip invocation support requires adding the `Switch.System.Windows.Forms.UseLegacyToolTipDisplay=false` line to the AppContextSwitchOverrides value:
+`<AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false;Switch.UseLegacyAccessibilityFeatures.3=false" />`
+
+Note that enabling this feature requires opting in to the aforementioned accessibility features of .NET Framework 4.7.1 - 4.8.
```
I would add that if any of the accessibility features are not opted in but the tooltip display feature is opted in, users would see a runtime exception on the first access to these features, most likely on application start up. Please specify exception type here.

M-Lipin replied:
Thank you, I've added a note about enabling keyboard tooltip feature as you suggested. Thanks!
